hswidget/ds1widget/tile_records.go

```diff
@@ -0,0 +1,207 @@
+		Sequence:    byte(state.addFloorShadowState.sequence),
+		Unknown1:    byte(state.addFloorShadowState.unknown1),
+		Style:       byte(state.addFloorShadowState.style),
+		Unknown2:    byte(state.addFloorShadowState.unknown2),
+		HiddenBytes: byte(state.addFloorShadowState.hidden),
+	}
+
+	return newFloorShadowRecord
+}
+
+func (p *DS1Widget) createWallRecord() d2ds1.WallRecord {
+	state := p.getState()
+
+	newWall := d2ds1.WallRecord{
+		Type:        d2enum.TileType(state.addWallState.tileType),
+		Zero:        byte(state.addWallState.zero),
+		Prop1:       byte(state.addWallState.prop1),
+		Sequence:    byte(state.addWallState.sequence),
+		Unknown1:    byte(state.addWallState.unknown1),
+		Style:       byte(state.addWallState.style),
+		Unknown2:    byte(state.addWallState.unknown2),
+		HiddenBytes: byte(state.addWallState.hidden),
+	}
+
+	return newWall
+}
+
+func (p *DS1Widget) addFloor() {
+	state := p.getState()
+
+	state.addFloorShadowState.cb = func() {
+		newFloor := p.createFloorShadowRecord()
+
+		for y := range p.ds1.Tiles {
+			for x := range p.ds1.Tiles[y] {
+				p.ds1.Tiles[y][x].Floors = append(p.ds1.Tiles[y][x].Floors, newFloor)
+			}
+		}
+
+		p.ds1.NumberOfFloors++
+
+		p.recreateLayerStreamTypes()
+	}
+
+	state.mode = ds1EditorModeAddFloorShadow
+}
+
+func (p *DS1Widget) editFloor() {
+	state := p.getState()
+
+	state.addFloorShadowState.cb = func() {
+		newFloor := p.createFloorShadowRecord()
+
+		p.ds1.Tiles[state.tileY][state.tileY].Floors[state.object] = newFloor
+	}
+	state.mode = ds1EditorModeAddFloorShadow
+}
+
+func (p *DS1Widget) deleteFloorRecord() {
+	state := p.getState()
+
+	for y := range p.ds1.Tiles {
+		for x := range p.ds1.Tiles[y] {
+			newFloors := make([]d2ds1.FloorShadowRecord, 0)
+
+			for n, i := range p.ds1.Tiles[y][x].Floors {
+				if n != int(state.object) {
+					newFloors = append(newFloors, i)
+				}
+			}
+
+			p.ds1.Tiles[y][x].Floors = newFloors
+		}
+	}
+
+	p.ds1.NumberOfFloors--
+	p.recreateLayerStreamTypes()
+}
+
+func (p *DS1Widget) addWall() {
+	state := p.getState()
+
+	state.addWallState.cb = func() {
+		newWall := p.createWallRecord()
+
+		for y := range p.ds1.Tiles {
+			for x := range p.ds1.Tiles[y] {
+				p.ds1.Tiles[y][x].Walls = append(p.ds1.Tiles[y][x].Walls, newWall)
+			}
+		}
+
+		p.ds1.NumberOfWalls++
+
+		p.recreateLayerStreamTypes()
+	}
+
+	state.mode = ds1EditorModeAddWall
+}
+
+func (p *DS1Widget) editWall() {
+	state := p.getState()
+
+	state.addWallState.cb = func() {
+		newWall := p.createWallRecord()
+
+		p.ds1.Tiles[state.tileY][state.tileY].Walls[state.object] = newWall
+	}
+	state.mode = ds1EditorModeAddWall
+}
+
+func (p *DS1Widget) deleteWall() {
+	state := p.getState()
+
+	for y := range p.ds1.Tiles {
+		for x := range p.ds1.Tiles[y] {
+			newWalls := make([]d2ds1.WallRecord, 0)
+
+			for n, i := range p.ds1.Tiles[y][x].Walls {
+				if n != int(state.object) {
+					newWalls = append(newWalls, i)
+				}
+			}
```

Comment:
probably should change `n` and `i` to something more descriptive

Reply:
well, n (natural number) is ok, but `i` could be replaced

Reply:
yeah, i'm okay with that. I probably wrote that code to begin with :x